lib/util/Components.js

```diff
@@ -581,45 +581,62 @@ function componentRule(rule, context) {
      * @returns {ASTNode} component node, null if we are not in a component
      */
     getParentStatelessComponent() {
-      let scope = context.getScope();
-      while (scope) {
-        const node = scope.block;
-        const isFunction = /Function/.test(node.type); // Functions
-        const isArrowFunction = astUtil.isArrowFunction(node);
-        const enclosingScope = isArrowFunction ? utils.getArrowFunctionScope(scope) : scope;
-        const enclosingScopeParent = enclosingScope && enclosingScope.block.parent;
-        const isClass = enclosingScope && astUtil.isClass(enclosingScope.block);
-        const isMethod = enclosingScopeParent && enclosingScopeParent.type === 'MethodDefinition'; // Classes methods
-        const isArgument = node.parent && node.parent.type === 'CallExpression'; // Arguments (callback, etc.)
-        // Attribute Expressions inside JSX Elements (<button onClick={() => props.handleClick()}></button>)
-        const isJSXExpressionContainer = node.parent && node.parent.type === 'JSXExpressionContainer';
-        const pragmaComponentWrapper = this.getPragmaComponentWrapper(node);
-        if (isFunction && pragmaComponentWrapper) {
-          return pragmaComponentWrapper;
+      /**
+       * @param {ASTNode} node
+       * @returns {boolean}
+       */
+      function isInAllowedPositiionForComponent(node) {
+        switch (node.parent.type) {
+          case 'VariableDeclarator':
+          case 'AssignmentExpression':
+          case 'Property':
+          case 'ReturnStatement':
+          case 'ExportDefaultDeclaration': {
+            return true;
+          }
+          case 'SequenceExpression': {
+            return isInAllowedPositiionForComponent(node.parent) &&
+              node === node.parent.expressions[node.parent.expressions.length - 1];
+          }
+          default:
+            return false;
         }
-        // Stop moving up if we reach a class or an argument (like a callback)
-        if (isClass || isArgument) {
-          return null;
+      }
+
+      /**
+       * Get node if node is a stateless component, or node.parent in cases like
+       * `React.memo` or `React.forwardRef`. Otherwise returns `undefined`.
+       * @param {ASTNode} node
+       * @returns {ASTNode | undefined}
+       */
+      function getStatelessComponent(node) {
```

Comment:
this function should also be hoisted up to module level

Reply:
I 100% prefer hoisted functions, but in this case, hoisting up this function will move it very far away from its usage (because the functions are in a giant `utils` object literal), which I do not feel good about.

Reply:
Perhaps i shouldn't have used the word "hoisted" :-) i don't like hoisting, but i think functions should be defined as far away as possible to maximize reuse.

Reply:
Hoisting is a word with overloaded meanings. I mean `hoisted` as `not nested function`. 

I feel functions dealing with the same topic should be co-located close together, so that reader do not need to jump around to understand a functionality.

Reply:
I think that's fine when the two positions are equivalent; in this case it will impact performance since it will be creating a new function many times, instead of reusing the same one from before.